packages/maskbook/src/web3/hooks/useERC20TokenApproveCallback.ts

```diff
@@ -34,53 +38,62 @@ export function useERC20TokenApproveCallback(address: string, amount?: string, s
-                    {
-                        gas: estimatedGas,
-                        from: account,
-                        to: erc20Contract.options.address,
-                    },
-                    (error, hash) => {
-                        if (error) reject(error)
-                        else {
-                            resolve(hash)
-                            setApproveHash(hash)
-                        }
-                    },
-                )
-            })
+            if (resetState !== ApproveState.APPROVED) {
+                await resetUSDT()
+            }
+            const to = erc20Contract.options.address
+            let approve: TransactionObject<boolean>
+            let estimatedGas: number
+            try {
+                approve = erc20Contract.methods.approve(spender, useExact ? amount : MaxUint256)
+                estimatedGas = await approve.estimateGas({ from, to })
+            } catch {
+                approve = erc20Contract.methods.approve(spender, amount)
+                estimatedGas = await approve.estimateGas({ from, to })
+            }
+            const { transactionHash } = await approve.send({ gas: estimatedGas, from, to })
+            setApproveHash(transactionHash)
+            return transactionHash
         },
-        [approveState, amount, account, spender, erc20Contract],
+        [approveState, resetState, resetUSDT, amount, from, spender, erc20Contract],
     )
 
     const resetCallback = useCallback(() => {
         setApproveHash('')
         revalidateBalance()
         revalidateAllowance()
-    }, [])
+    }, [revalidateBalance, revalidateAllowance])
 
     return [approveState, approveCallback, resetCallback] as const
 }
+
+function useERC20TokenUSDTResetCallback(address: string, spender?: string) {
+    const from = useAccount()
+    const usdtAddress = useConstant(CONSTANTS, 'USDT_ADDRESS')
+    const erc20Contract = useERC20TokenContract(address === usdtAddress ? address : '')
+
+    const [approveHash, setApproveHash] = useState('')
+    const receipt = useTransactionReceipt(approveHash)
+
+    const approveState = useMemo(
+        () => (receipt?.blockHash || erc20Contract === null ? ApproveState.APPROVED : ApproveState.NOT_APPROVED),
+        [erc20Contract, receipt?.blockHash],
+    )
+
+    const approveCallback = useCallback(async () => {
+        if (approveState !== ApproveState.NOT_APPROVED) return
+        if (!spender || !erc20Contract) return
+        const approve = erc20Contract.methods.approve(spender, '0')
+        const to = erc20Contract.options.address
+        const estimatedGas = await approve.estimateGas({ from, to })
+        const { transactionHash } = await approve.send({ gas: estimatedGas, from, to })
+        setApproveHash(transactionHash)
+        return transactionHash
+    }, [approveState, from, spender, erc20Contract])
+
+    return [approveState, approveCallback] as const
+}
```

Comment:
shoud add a reset callback for resetting receipt and transaction hash.

Reply:
1. the is internal hook
2. unused `useERC20TokenBalance` and `useERC20TokenAllowance`.